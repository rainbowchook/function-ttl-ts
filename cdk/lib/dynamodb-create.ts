import { Construct } from 'constructs'
import {
  Table,
  ITable,
  AttributeType,
  BillingMode,
  StreamViewType,
} from 'aws-cdk-lib/aws-dynamodb'
import { RemovalPolicy } from 'aws-cdk-lib'

export class DynamoDB extends Construct {
  public readonly table: ITable
  constructor(scope: Construct, id: string) {
    super(scope, id)
    const tableName = 'TTLTable-v3'
    const table = new Table(this, 'TestDynamoDBTable', {
      tableName,
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE,
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
      timeToLiveAttribute: 'ttl',
    })

    this.table = table
  }
}
