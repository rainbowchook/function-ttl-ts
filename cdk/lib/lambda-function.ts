import * as path from 'node:path'

import { Construct } from 'constructs'
import { Role } from 'aws-cdk-lib/aws-iam'
import { Function, Runtime, Code, Architecture } from 'aws-cdk-lib/aws-lambda'
import { Duration, RemovalPolicy } from 'aws-cdk-lib'
import { ITable } from 'aws-cdk-lib/aws-dynamodb'
import { Bucket } from 'aws-cdk-lib/aws-s3'
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs'

export class LambdaFn extends Construct {
  public readonly lambdaFunction: Function
  constructor(
    scope: Construct,
    id: string,
    resources: { role: Role; table: ITable; s3Bucket: Bucket }
  ) {
    super(scope, id)

    const functionName = 'ProcessDynamoDBTTLRecords'
    const lambdaFunction = new Function(this, 'Lambda', {
      functionName,
      runtime: Runtime.NODEJS_16_X,
      code: Code.fromAsset(path.join(__dirname, '../../lambda/build')),
      handler: 'index.handler',
      // runtime: Runtime.PROVIDED_AL2023,
      // code: Code.fromAsset(path.join(__dirname, '../cmd/lambda/bootstrap.zip')),
      // architecture: Architecture.X86_64,
      // handler: 'bootstrap',
      role: resources.role,
      timeout: Duration.seconds(60),
      environment: {
        DYNAMODB_TABLE_NAME: resources.table.tableName,
        BUCKET_NAME: resources.s3Bucket.bucketName,
      },
    })

    this.lambdaFunction = lambdaFunction
  }
}
