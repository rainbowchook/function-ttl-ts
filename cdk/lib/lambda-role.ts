import { Construct } from "constructs";
import { ManagedPolicy, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";

export class LambdaRole extends Construct {
  public readonly role
  constructor(scope: Construct, id: string) {
    super(scope, id)

    const role = new Role(this, 'LambdaRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      ]
    })

    this.role = role
  }
}