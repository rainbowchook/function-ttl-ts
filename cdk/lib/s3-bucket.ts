import { Construct } from 'constructs'
import { Bucket, BucketEncryption } from 'aws-cdk-lib/aws-s3'
import { RemovalPolicy } from 'aws-cdk-lib'

export class ResultsS3Bucket extends Construct {
  public readonly s3Bucket
  constructor(scope: Construct, id: string) {
    super(scope, id)
    const bucketName = 'report-bucket-v2'
    const s3Bucket = new Bucket(this, 'ProjectS3Bucket', {
      bucketName,
      eventBridgeEnabled: true,
      removalPolicy: RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE,
      versioned: false,
      encryption: BucketEncryption.S3_MANAGED,
    })

    this.s3Bucket = s3Bucket
  }
}
