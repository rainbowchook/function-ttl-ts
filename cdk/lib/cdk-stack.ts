import { CfnOutput, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib'
import { Construct } from 'constructs'
import { DynamoDB } from './dynamodb-create'
import { LambdaRole } from './lambda-role'
import { LambdaFn } from './lambda-function'
import { Alarm, TreatMissingData } from 'aws-cdk-lib/aws-cloudwatch'
import { ResultsS3Bucket } from './s3-bucket'
import { FilterPattern, LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs'
import { LambdaDestination } from 'aws-cdk-lib/aws-logs-destinations'

export class FunctionTTLProcessingStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props)

    // create the DynamoDB table with TTL
    const { table } = new DynamoDB(this, 'TestDynamoDB')

    // create the S3 bucket to put TTL processing result
    const { s3Bucket } = new ResultsS3Bucket(this, 'ReportS3Bucket')

    // create the lambda service execution role
    const { role } = new LambdaRole(this, 'LambdaRole', { s3Bucket })

    // create the lambda
    const { lambdaFunction } = new LambdaFn(this, 'LambdaFunction', {
      role,
      table,
      s3Bucket
    })

    // grant permissions to principal to DescribeStream, GetRecords, GetShardIterator, ListStreams
    table.grantStreamRead(lambdaFunction)

    // create the CloudWatch alarm on lambda timeout
    if (lambdaFunction.timeout) {
      new Alarm(this, 'NewAlarm', {
        metric: lambdaFunction.metricDuration().with({
          statistic: 'Maximum',
        }),
        evaluationPeriods: 1,
        datapointsToAlarm: 1,
        threshold: lambdaFunction.timeout.toMilliseconds(),
        treatMissingData: TreatMissingData.IGNORE,
        alarmName: 'lambda-timeout-alarm',
      })
    }
    
    // create log group and indicate retention period
    const lambdaLogGroup = new LogGroup(this, 'TTLLogGroup', {
      logGroupName: `/aws/lambda/platform/${lambdaFunction.functionName}`,
      retention: 5 as RetentionDays,
      removalPolicy: RemovalPolicy.DESTROY
    })
    lambdaLogGroup.addSubscriptionFilter('LambdaLogGroupSubscription', {
      destination: new LambdaDestination(lambdaFunction),
      filterPattern: FilterPattern.allEvents()
    })

    // output the Lambda function name
    new CfnOutput(this, 'TTLLambdaOutput', {
      value: lambdaFunction.functionArn
    })
  }
}
