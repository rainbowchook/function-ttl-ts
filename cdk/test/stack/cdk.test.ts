import * as cdk from 'aws-cdk-lib'
import { Capture, Match, Template } from 'aws-cdk-lib/assertions'
import * as Cdk from '../../lib/cdk-stack'
import { Runtime } from 'aws-cdk-lib/aws-lambda'
import { BillingMode, StreamViewType } from 'aws-cdk-lib/aws-dynamodb'

// Instantiate the cdk app
const app = new cdk.App()

// create stack
const stack = new Cdk.FunctionTTLProcessingStack(app, 'CdkLambdaStack')

// prepare the stack for assertions
const template = Template.fromStack(stack)

describe('FunctionTTLProcessingStack resources created', () => {
  // Snapshot testing
  test('Resources created matches the snapshot', () => {
    // match with snapshot
    expect(template.toJSON()).toMatchSnapshot()
  })

  //Fine-grained testing
  test('Specified resources created', () => {
    template.resourceCountIs('AWS::DynamoDB::Table', 1)
    template.resourceCountIs('AWS::S3::Bucket', 1)
    template.resourcePropertiesCountIs(
      'AWS::Lambda::Function',
      { Runtime: Runtime.NODEJS_16_X.name, Environment: Match.anyValue() },
      1
    )

    template.hasResourceProperties('AWS::DynamoDB::Table', {})
    template.hasResourceProperties('AWS::S3::Bucket', {})
    template.hasResourceProperties('Custom::S3BucketNotifications', {})
    template.hasResourceProperties('AWS::IAM::Role', {})
    template.hasResourceProperties('AWS::IAM::Policy', {})
    template.hasResourceProperties('AWS::Lambda::Function', {})
    template.hasResourceProperties('Custom::LogRetention', {})
    template.hasResourceProperties('AWS::CloudWatch::Alarm', {})
  })
})

describe('DynamoDB tests', () => {
  test('DynamoDB table has correct properties', () => {
    const ttlEnableCapture = new Capture()
    template.hasResourceProperties('AWS::DynamoDB::Table', {
      BillingMode: BillingMode.PAY_PER_REQUEST,
      StreamSpecification: {
        StreamViewType: StreamViewType.NEW_AND_OLD_IMAGES,
      },
      TimeToLiveSpecification: {
        AttributeName: 'ttl',
        Enabled: ttlEnableCapture,
      },
    })

    expect(ttlEnableCapture.asBoolean()).toBe<boolean>(true)
  })
})

describe('Lambda tests', () => {
  test('Lambda function has correct properties', () => {
    console.log(template.findResources('AWS::DynamoDB::Table', {}))

    const dependencyCapture = new Capture()
    template.hasResource('AWS::Lambda::Function', {
      Properties: {
        Code: {
          S3Bucket: {
            'Fn::Sub': Match.stringLikeRegexp('^cdk-.*?-assets-.*?$'),
          },
        },
        Environment: {
          Variables: {
            NODE_OPTIONS: '--enable-source-maps',
            DYNAMODB_TABLE_NAME: Match.anyValue(),
            BUCKET_NAME: Match.anyValue(),
          },
        },
        Handler: 'index.handler',
        Runtime: Runtime.NODEJS_16_X.name,
        Timeout: 60,
      },
      DependsOn: dependencyCapture,
    })
    expect(dependencyCapture).not.toBeNull()
    expect(dependencyCapture.asArray().length).toBeGreaterThanOrEqual(1)
  })

  test('Lambda function execution role has correct IAM permissions', () => {
    const roleCapture = new Capture()
    const actionCapture = new Capture()
    template.hasResourceProperties('AWS::IAM::Role', {
      AssumeRolePolicyDocument: Match.objectLike({
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
          },
        ],
        Version: '2012-10-17',
      }),
      ManagedPolicyArns: [
        {
          'Fn::Join': Match.arrayWith([
            ['arn:', { Ref: 'AWS::Partition' }, roleCapture],
          ]),
        },
      ],
      Policies: Match.arrayWith([
        Match.objectLike({
          PolicyDocument: Match.objectLike({
            Statement: [
              {
                Action: actionCapture,
                Effect: 'Allow',
                Resource: {
                  'Fn::Join': Match.arrayWith([
                    [
                      {
                        'Fn::GetAtt': Match.arrayWith([
                          Match.stringLikeRegexp('^.*?Bucket.*?$'),
                        ]),
                      },
                      '/*',
                    ],
                  ]),
                },
              },
            ],
            Version: '2012-10-17',
          }),
        }),
      ]),
    })

    expect(
      roleCapture.asString().match(/AWSLambdaBasicExecutionRole/)
    ).not.toBeNull()
    expect(actionCapture.asString()).toMatch(/s3:PutObject/)
  })

  test('Lambda has correct event source mapping', () => {
    template.hasResourceProperties('AWS::Lambda::EventSourceMapping', {
      FilterCriteria: {
        Filters: [
          {
            Pattern: Match.serializedJson({
              userIdentity: {
                type: Match.arrayEquals(['service']),
                principalId: Match.arrayEquals(['dynamodb.amazonaws.com']),
              },
            }),
          },
        ],
      },
    })
  })

  test('Lambda not running in VPC', () => {
    template.hasResource('AWS::Lambda::Function', {
      Vpc: Match.absent(),
    })
  })
})

describe('S3 tests', () => {})

describe('CloudWatch tests', () => {})
