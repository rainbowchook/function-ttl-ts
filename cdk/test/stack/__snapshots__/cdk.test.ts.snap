// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FunctionTTLProcessingStack resources created Resources created matches the snapshot 1`] = `
{
  "Outputs": {
    "DynamoDBTableName": {
      "Description": "DynamoDB Table with TTL attribute",
      "Value": {
        "Ref": "TestDynamoDBTestDynamoDBTable67588611",
      },
    },
    "S3BucketName": {
      "Description": "S3 Bucket name",
      "Value": {
        "Ref": "ReportS3BucketProjectS3Bucket985029D9",
      },
    },
    "TTLProcessingLambdaOutput": {
      "Value": {
        "Fn::GetAtt": [
          "LambdaFunctionLambda4835DFE3",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691": {
      "DependsOn": [
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "import boto3  # type: ignore
import json
import logging
import urllib.request

s3 = boto3.client("s3")

EVENTBRIDGE_CONFIGURATION = 'EventBridgeConfiguration'

CONFIGURATION_TYPES = ["TopicConfigurations", "QueueConfigurations", "LambdaFunctionConfigurations"]

def handler(event: dict, context):
  response_status = "SUCCESS"
  error_message = ""
  try:
    props = event["ResourceProperties"]
    bucket = props["BucketName"]
    notification_configuration = props["NotificationConfiguration"]
    request_type = event["RequestType"]
    managed = props.get('Managed', 'true').lower() == 'true'
    stack_id = event['StackId']

    if managed:
      config = handle_managed(request_type, notification_configuration)
    else:
      config = handle_unmanaged(bucket, stack_id, request_type, notification_configuration)

    put_bucket_notification_configuration(bucket, config)
  except Exception as e:
    logging.exception("Failed to put bucket notification configuration")
    response_status = "FAILED"
    error_message = f"Error: {str(e)}. "
  finally:
    submit_response(event, context, response_status, error_message)

def handle_managed(request_type, notification_configuration):
  if request_type == 'Delete':
    return {}
  return notification_configuration

def handle_unmanaged(bucket, stack_id, request_type, notification_configuration):
  external_notifications = find_external_notifications(bucket, stack_id)

  if request_type == 'Delete':
    return external_notifications

  def with_id(notification):
    notification['Id'] = f"{stack_id}-{hash(json.dumps(notification, sort_keys=True))}"
    return notification

  notifications = {}
  for t in CONFIGURATION_TYPES:
    external = external_notifications.get(t, [])
    incoming = [with_id(n) for n in notification_configuration.get(t, [])]
    notifications[t] = external + incoming

  if EVENTBRIDGE_CONFIGURATION in notification_configuration:
    notifications[EVENTBRIDGE_CONFIGURATION] = notification_configuration[EVENTBRIDGE_CONFIGURATION]
  elif EVENTBRIDGE_CONFIGURATION in external_notifications:
    notifications[EVENTBRIDGE_CONFIGURATION] = external_notifications[EVENTBRIDGE_CONFIGURATION]

  return notifications

def find_external_notifications(bucket, stack_id):
  existing_notifications = get_bucket_notification_configuration(bucket)
  external_notifications = {}
  for t in CONFIGURATION_TYPES:
    external_notifications[t] = [n for n in existing_notifications.get(t, []) if not n['Id'].startswith(f"{stack_id}-")]

  if EVENTBRIDGE_CONFIGURATION in existing_notifications:
    external_notifications[EVENTBRIDGE_CONFIGURATION] = existing_notifications[EVENTBRIDGE_CONFIGURATION]

  return external_notifications

def get_bucket_notification_configuration(bucket):
  return s3.get_bucket_notification_configuration(Bucket=bucket)

def put_bucket_notification_configuration(bucket, notification_configuration):
  s3.put_bucket_notification_configuration(Bucket=bucket, NotificationConfiguration=notification_configuration)

def submit_response(event: dict, context, response_status: str, error_message: str):
  response_body = json.dumps(
    {
      "Status": response_status,
      "Reason": f"{error_message}See the details in CloudWatch Log Stream: {context.log_stream_name}",
      "PhysicalResourceId": event.get("PhysicalResourceId") or event["LogicalResourceId"],
      "StackId": event["StackId"],
      "RequestId": event["RequestId"],
      "LogicalResourceId": event["LogicalResourceId"],
      "NoEcho": False,
    }
  ).encode("utf-8")
  headers = {"content-type": "", "content-length": str(len(response_body))}
  try:
    req = urllib.request.Request(url=event["ResponseURL"], headers=headers, data=response_body, method="PUT")
    with urllib.request.urlopen(req) as response:
      print(response.read().decode("utf-8"))
    print("Status code: " + response.reason)
  except Exception as e:
      print("send(..) failed executing request.urlopen(..): " + str(e))
",
        },
        "Description": "AWS CloudFormation handler for "Custom::S3BucketNotifications" resources (@aws-cdk/aws-s3)",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": [
          {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaFunctionLambda4835DFE3": {
      "DependsOn": [
        "LambdaRoleDefaultPolicyE834987C",
        "LambdaRole64EBED7A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b4df8631b33066548afe43bb7b47bcc952a4448db762d1d86831f211e3fe90d7.zip",
        },
        "Description": "DynamoDB TTL Processing Lambda",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": {
              "Ref": "ReportS3BucketProjectS3Bucket985029D9",
            },
            "DYNAMODB_TABLE_NAME": {
              "Ref": "TestDynamoDBTestDynamoDBTable67588611",
            },
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "FunctionName": "ProcessDynamoDBTTLRecords",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LambdaRole64EBED7A",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaFunctionLambdaDynamoDBEventSourceCdkLambdaStackTestDynamoDBTestDynamoDBTableDA72C4EE20372877": {
      "Properties": {
        "BatchSize": 5,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "TestDynamoDBTestDynamoDBTable67588611",
            "StreamArn",
          ],
        },
        "FilterCriteria": {
          "Filters": [
            {
              "Pattern": "{"userIdentity":{"type":["service"],"principalId":["dynamodb.amazonaws.com"]}}",
            },
          ],
        },
        "FunctionName": {
          "Ref": "LambdaFunctionLambda4835DFE3",
        },
        "MaximumRetryAttempts": 10,
        "StartingPosition": "TRIM_HORIZON",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "LambdaFunctionLambdaLogRetention6CDD8563": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "LambdaFunctionLambda4835DFE3",
              },
            ],
          ],
        },
        "RetentionInDays": 5,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaRole64EBED7A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:PutObject",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        {
                          "Fn::GetAtt": [
                            "ReportS3BucketProjectS3Bucket985029D9",
                            "Arn",
                          ],
                        },
                        "/*",
                      ],
                    ],
                  },
                  "Sid": "CustomS3Policy",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "s3",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaRoleDefaultPolicyE834987C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TestDynamoDBTestDynamoDBTable67588611",
                  "StreamArn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaRoleDefaultPolicyE834987C",
        "Roles": [
          {
            "Ref": "LambdaRole64EBED7A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5bc602ecde93c947efe5899ae355f999986a1acbe610b1c0b9c468d738857555.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "NewAlarmD2C7D5DE": {
      "Properties": {
        "AlarmName": "lambda-timeout-alarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "Dimensions": [
          {
            "Name": "FunctionName",
            "Value": {
              "Ref": "LambdaFunctionLambda4835DFE3",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Duration",
        "Namespace": "AWS/Lambda",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 60000,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ReportS3BucketProjectS3Bucket985029D9": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "test-report-bucket-v3",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "ReportS3BucketProjectS3BucketNotifications97482F46": {
      "Properties": {
        "BucketName": {
          "Ref": "ReportS3BucketProjectS3Bucket985029D9",
        },
        "Managed": true,
        "NotificationConfiguration": {
          "EventBridgeConfiguration": {},
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3BucketNotifications",
    },
    "TestDynamoDBTestDynamoDBTable67588611": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "StreamSpecification": {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableName": "TTLTable-v3",
        "TimeToLiveSpecification": {
          "AttributeName": "ttl",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
